ocfweb/account/vhost_mail.py

```diff
@@ -117,6 +135,137 @@ def vhost_mail_update(request):
+    vhost = _get_vhost(user, domain)
+    if not vhost:
+        _error(request, 'You cannot use the domain: "{}"'.format(domain))
+
+    with _txn() as c:
+        addresses = (
+            addr for addr in sorted(vhost.get_forwarding_addresses(c))
+            if not addr.is_wildcard
+        )
+
+    response = HttpResponse(
+        _write_csv(addresses),
+        content_type='text/csv',
+    )
+    # See https://docs.djangoproject.com/en/1.11/ref/request-response/
+    response['Content-disposition'] = 'attachment; filename="{}.csv"'.format(domain)
+    return response
+
+
+@login_required
+@group_account_required
+@require_POST
+def vhost_mail_csv_import(request, domain):
+    user = logged_in_user(request)
+    vhost = _get_vhost(user, domain)
+    if not vhost:
+        _error(request, 'You cannot use the domain: "{}"'.format(domain))
+
+    addresses = _parse_csv(request, domain)
+
+    # Add new addresses and update existing if the recipients changed
+    with _txn() as c:
+        existing_addrs = {
+            addr.address: addr
+            for addr in vhost.get_forwarding_addresses(c)
+        }
+
+        for from_addr, to_addrs in addresses.items():
+            if from_addr in existing_addrs:
+                existing_addr = existing_addrs[from_addr]
+                if to_addrs != existing_addr.forward_to:
+                    new = MailForwardingAddress(
+                        address=from_addr,
+                        crypt_password=existing_addr.crypt_password,
+                        forward_to=to_addrs,
+                        last_updated=None,
+                    )
+                    vhost.remove_forwarding_address(c, from_addr)
+                else:
+                    new = None
+            else:
+                new = MailForwardingAddress(
+                    address=from_addr,
+                    crypt_password=None,
+                    forward_to=to_addrs,
+                    last_updated=None,
+                )
+
+            if new is not None:
+                vhost.add_forwarding_address(c, new)
+
+    messages.add_message(request, messages.SUCCESS, 'Import successful!')
+    return _redirect_back()
+
+
+def _write_csv(addresses):
+    """Turn a collection of vhost forwarding addresses into a CSV
+    string for user download."""
+    buf = io.StringIO()
+    writer = csv.writer(buf)
+    for addr in addresses:
+        writer.writerow((
+            addr.address.split('@')[0],
+            ' '.join(sorted(addr.forward_to)),
+        ))
+
+    return buf.getvalue()
+
+
+def _parse_csv(request, domain):
+    """Parse, validate, and return addresses from the file uploaded
+    with the CSV upload button/form."""
+    csv_file = request.FILES.get('csv_file')
+    if not csv_file:
+        _error(request, 'Missing CSV file!')
+
+    addresses = {}
+    with io.TextIOWrapper(csv_file, encoding='utf-8') as f:
+        reader = csv.reader(f)
+        for i, row in enumerate(reader):
```

Comment:
I don't know if we should be concerned about somebody trying to upload an inanely large CSV...

Reply:
I originally thought about this briefly and found out that our nginx is configured to put a 10M limit on uploaded files, so no one can upload a truly enormous file. Given that there are much worse ways you could DoS the OCF, I don't think anyone is likely to abuse this anyways. Still, 10M is enough to upload as much as a million lines of CSV.

As far as the website goes, this could stall a web worker for a few minutes as it parses the file and makes a zillion SQL queries. If you did it 100 times, you might be able to DoS the website. More dangerously, if you uploaded enough unique addresses, you could flood the mail vhost table which might break postfix, which is bad.

Do you suggest we do anything about it? If I were serious about it, I would set a line limit on uploaded CSVs and a hard limit in ocflib on the number of addresses for a single vhost, which would prevent flooding the table.

Reply:
OK, I've addressed the other concerns, and I don't think this needs to stall the PR. If we want to harden our vhost mail database, we can make a ticket for it or an issue in ocflib. This feature is probably fine to release as-is.